### PR TITLE
Python: promote XPath injection query

### PR DESCRIPTION
**Pull request by yoff**
Commits are structured in an attempt to preserve the original files before they get adapted.

Possible concerns:
 - the precision was never tested and no sanitizers are implemented (but js and java also seem to not have sanitizers)
 - should we model more of lxml? XPathEvaluator, soupparser
 - should we model more frameworks?
   - `stdlib` has `xml.etree.ElementTree`
   - [py-dom-xpath](https://github.com/neild/py-dom-xpath)
   - PyXML
   - [parsel](https://github.com/scrapy/parsel)

### CODE REVIEW

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Very nice commit structure :muscle: 

To answer your questions:

> the precision was never tested and no sanitizers are implemented (but js and java also seem to not have sanitizers)

I looked back through the archives, and the precision was tested, but not determined since it didn't actually find any results. We should try running the final query against all of LGTM, to see what the status is now.

> should we model more of lxml? XPathEvaluator, soupparser

I had to look it up ([here](https://lxml.de/apidoc/lxml.etree.html#lxml.etree.XPathEvaluator)), and yes, that sounds like it's easy to extend with this functionality... let's do it :+1:

> should we model more frameworks?

Would be cool with support for standard library :+1:

From reading https://docs.python.org/3.10/library/xml.etree.elementtree.html#elementtree-xpath it should also be quite simple. I think we can utilize some of the modeling from the [XXE PR](https://github.com/github/codeql/pull/6112) as well, once that is promoted. So for now, I think it's ok for the "parsed document" to be fairly simple for stdlib etree :+1:

File: python/ql/src/Security/CWE-643/XpathInjection.ql [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/**
+ * @name XPath query built from user-controlled sources
+ * @description Building a XPath query from user-controlled sources is vulnerable to insertion of
+ *              malicious Xpath code by the user.
+ * @kind path-problem
+ * @problem.severity error
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/Security/CWE-643/XpathInjection.ql:6]: Thanks!

File: python/ql/lib/semmle/python/frameworks/Lxml.qll [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            .getMember(["parse", "fromstring", "fromstringlist", "HTML", "XML"])
+            .getReturn()
+            .getMember("xpath")
+            .getACall()
+    }
+
+    override DataFlow::Node getXPath() { result in [this.getArg(0), this.getArgByName("_path")] }
+
+    // TODO: implement when we get call nodes
+    override DataFlow::Node getTree() { none() }
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Lxml.qll:73]: Indeed, we can easily add it when the need arises. Then we just need to go through all supported frameworks and implement it, but the compiler will tell us where.

File: python/ql/lib/semmle/python/frameworks/Lxml.qll [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+   * - https://lxml.de/apidoc/lxml.etree.html#lxml.etree.fromstringlist
+   * - https://lxml.de/apidoc/lxml.etree.html#lxml.etree.HTML
+   * - https://lxml.de/apidoc/lxml.etree.html#lxml.etree.XML
+   */
+  class XPathCall extends XPathExecution::Range, DataFlow::CallCfgNode {
+    XPathCall() {
+      this =
+        API::moduleImport("lxml")
+            .getMember("etree")
+            .getMember(["parse", "fromstring", "fromstringlist", "HTML", "XML"])
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Lxml.qll:63]: Sounds good.

**Comment by RasmusWL** [comment]
@yoff I would actually also recommend that we make a `XML` module in the Concepts file, since I know that we will get more XML stuff when we promote the XXE query, so otherwise we will flood the top-level namespace with XML stuff :grimacing:

File: python/ql/test/library-tests/frameworks/stdlib/XPathExecution.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import xml.etree.ElementTree as ET
+tree = ET.parse('country_data.xml')
+root = tree.getroot()
+
+root.find(match, namespaces=ns)  # $ getXPath=match
+root.findall(match, namespaces=ns)  # $ getXPath=match
+root.findtext(match, default=None, namespaces=ns)  # $ getXPath=match
+
+from xml.etree.ElementTree import ElementTree
+tree = ElementTree()
```
Comment (RasmusWL) [inline python/ql/test/library-tests/frameworks/stdlib/XPathExecution.py:13]: NIT

```suggestion
tree = ET.ElementTree()
```

**Review by RasmusWL — COMMENTED** [review]
Overall LGTM!

However, I did not see any comments about the precision. Did you confirm that `high` is the right choice?

**Comment by yoff** [comment]
> However, I did not see any comments about the precision. Did you confirm that `high` is the right choice?
Yes, you should see a mention from an issue with a link to a full run. Looking through the results there, I found it to be fairly precise.

File: python/ql/test/library-tests/frameworks/stdlib/XPathExecution.py (hunk 4 above)
Comment (yoff) [reply python/ql/test/library-tests/frameworks/stdlib/XPathExecution.py]: Thanks

**Review by RasmusWL — APPROVED** [review]
> Yes, you should see a mention from an issue with a link to a full run. Looking through the results there, I found it to be fairly precise.

great :+1:

**Pull request merged**